stl/inc/syncstream

```diff
@@ -0,0 +1,339 @@
+// syncstream standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _SYNCSTREAM_
+#define _SYNCSTREAM_
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#ifndef __cpp_lib_syncbuf
+#pragma message("The contents of <syncstream> are available only with C++20 syncbuf support.")
+#else // ^^^ !defined(__cpp_lib_syncbuf) / defined(__cpp_lib_syncbuf) vvv
+#include <memory>
+#include <ostream>
+#include <shared_mutex>
+#include <streambuf>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#pragma push_macro("emit")
+#undef new
+#undef emit
+
+_STD_BEGIN
+
+extern "C" shared_mutex& __cdecl _Get_mutex_for_instance(void* _Ptr);
+extern "C" void __cdecl _Add_mutex_for_instance_or_increment(void* _Ptr) noexcept;
+extern "C" void __cdecl _Delete_mutex_for_instance_or_decrement(void* _Ptr) noexcept;
+
+// CLASS TEMPLATE basic_syncbuf
+template <class _Elem, class _Traits, class _Alloc>
+class basic_syncbuf : public basic_streambuf<_Elem, _Traits> {

```

Comment:
I am having a look at [P0753R2](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0753r2.pdf) `osyncstream` Manipulators , which I thought to be an addition to the implementation of [P0053R7](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0053r7.pdf) as the name implies, however, there's a scary note 😨.

```
To work around the issue that the Allocator template argument can not be deduced, implementations can
introduce an intermediate base class to basic_osyncbuf that manages its emit_on_sync flag
```
I am thinking about it now, but apparently, it will be required to create an intermediate class that only manages `_Emit_on_sync` member variable.

- What do you think? Could there be another workaround that I have missed?
- If an intermediate class is required, is there a naming convention for such a thing?
- Is it wise for the intermediate class to manage `_Emit_on_sync` variable only without interacting with `basic_streambuf`'s members at all?

Reply:
Can't imagine another workaround with equal efficiency. A base class with a virtual member function would be strictly worse.

There isn't much in the way of consistent precedent for base class naming - we have `_Vector_val` which holds its data members (erasing the allocator type), while `function` inherits from `_Func_class` (not a very descriptive name), `optional` has many base classes like `_Optional_construct_base`, etc.

`_Emit_on_sync_manager`, `_Emit_on_sync_base`, `_Basic_osyncbuf_impl`, etc. could all be plausible.